sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -359,6 +471,50 @@ class RocksDB(
     this
   }
 
+  /**
+   * Initialize in memory values based on the metadata loaded from DFS.
+   * @param metadata: metadata loaded from DFS
+   */
+  private def init(metadata: RocksDBCheckpointMetadata): Unit = {
+
+    setInitialCFInfo()
+    metadata.columnFamilyMapping.foreach { mapping =>
+      colFamilyNameToIdMap.putAll(mapping.asJava)
+    }
+
+    metadata.maxColumnFamilyId.foreach { maxId =>
+      maxColumnFamilyId.set(maxId)
+    }
+    openDB()
+    numKeysOnWritingVersion = if (!conf.trackTotalNumberOfRows) {
+      // we don't track the total number of rows - discard the number being track
+      -1L
+    } else if (metadata.numKeys < 0) {
+      // we track the total number of rows, but the snapshot doesn't have tracking number
+      // need to count keys now
+      countKeys()
+    } else {
+      metadata.numKeys
+    }
+  }
+
+  def load(
+      version: Long,
+      stateStoreCkptId: Option[String] = None,
+      readOnly: Boolean = false): RocksDB = {
+    assert(version >= 0)
+    acquire(LoadStore)
+    recordedMetrics = None
+    logInfo(log"Loading ${MDC(LogKeys.VERSION_NUM, version)} with stateStoreCkptId: ${
+      MDC(LogKeys.UUID, stateStoreCkptId.getOrElse(""))}")
+    if (enableStateStoreCheckpointIds) {
```

Comment:
Should be when stateStoreCkptId is non-empty, rather than `enableStateStoreCheckpointIds`. No matter whether we support V1->V2 compatibility or not, the source of truth of what we can load is from `stateStoreCkptId` loaded from the commit log, not the configuration, which is more about what version to be rewritten, nor read. Even if we don't support V1->V2, we still should use `stateStoreCkptId` but add check and explicitly fail the query, saying it is not supported. This is perhaps already done in driver?

Reply:
We also cannot do it here because in version 0 the snapshot unique id is None.

Reply:
Good point. Can we have a special treatment of version 0 here?

Reply:
sure